### PR TITLE
Added verification if LineString or MultiLineString patch shapes. Fix…

### LINKED ISSUE #40525: QGIS crashes when setting LineString patch shape to Polygon or Point
**Describe the bug**

QGIS crashes consistently when I try to set the patch shape of a line string to a polygon or point within a legend.

**How to Reproduce**

1. create New Temporary Scratch Layer of type LineString
2. create and open Layout
3. add legend item
4. uncheck Auto Update in the legend's Item Properties
5. double click New scratch layer and click the Legend Patch Shape property
6. enter text: `Polygon ((0 5, 10 5, 10 0, 0 0, 0 5))` **or** `Point (5 2)`

**QGIS and OS versions**

<pre>
QGIS version
3.16.1-Hannover
QGIS code revision
b381a90dca
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.1.4
Running against GDAL/OGR
3.1.4
Compiled against GEOS
3.8.1-CAPI-1.13.3
Running against GEOS
3.8.1-CAPI-1.13.3
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.3.2
Running against PROJ
Rel. 6.3.2, May 1st, 2020
OS Version
Windows 10 (10.0)
Active python plugins
AnotherDXF2Shape; 
FOEA; 
GroupStats; 
NNJoin; 
pathfinder; 
pluginbuilder3; 
plugin_reloader; 
SpreadsheetLayers; 
db_manager; 
MetaSearch; 
processing
</pre>
**Additional context**

## Report Details

**Crash ID**: [572f3a07ee109d1dc30475ca8b2f6b920fb6feda](https://github.com/qgis/QGIS/search?q=572f3a07ee109d1dc30475ca8b2f6b920fb6feda&type=Issues)

**Stack Trace**
<pre>
curveToPolygonF qgslegendpatchshape.cpp:79
QgsLegendPatchShape::toQPolygonF qgslegendpatchshape.cpp:156
QgsLineSymbolLayer::drawPreviewIcon qgssymbollayer.cpp:663
QgsSymbol::drawPreviewIcon qgssymbol.cpp:597
QgsSymbolLayerUtils::symbolPreviewPixmap qgssymbollayerutils.cpp:815
QgsSymbolLayerUtils::symbolPreviewIcon qgssymbollayerutils.cpp:764
QgsLegendPatchShapeButton::updatePreview qgslegendpatchshapebutton.cpp:283
QgsLegendPatchShapeButton::setShape qgslegendpatchshapebutton.cpp:166
QtPrivate::QFunctorSlotObject<<lambda_7ad474b84ba1cc2ff4010d9867dfc425>,0,QtPrivate::List<>,void>::impl qobjectdefs_impl.h:421
QMetaObject::activate :
QMetaObject::activate :
QMetaObject::activate :
QWidgetTextControl::qt_metacall :
QMetaObject::activate :
QTextDocumentPrivate::finishEdit :
QTextCursor::insertFragment :
QWidgetTextControl::insertFromMimeData :
QWidgetTextControl::paste :
QWidgetTextControl::isWordSelectionEnabled :
QWidgetTextControl::processEvent :
QWidgetTextControl::processEvent :
QPlainTextEdit::selectionChanged :
QPlainTextEdit::keyPressEvent :
QWidget::event :
QFrame::event :
QAbstractScrollArea::event :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify qgsapplication.cpp:458
QCoreApplication::notifyInternal2 :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify qgsapplication.cpp:458
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processKeyEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
UserCallWinProcCheckWow :
DispatchMessageWorker :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main main.cpp:1637
BaseThreadInitThunk :
RtlUserThreadStart :
</pre>

**QGIS Info**
QGIS Version: 3.16.1-Hannover
QGIS code revision: b381a90dca
Compiled against Qt: 5.11.2
Running against Qt: 5.11.2
Compiled against GDAL: 3.1.4
Running against GDAL: 3.1.4

**System Info**
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.18363

### PR DESCRIPTION
**Pull request by TurboGraphxBeige**
Fixes https://github.com/qgis/QGIS/issues/40525

Added a verification to make sure the patch shape is a LineString or MultiLineString to avoid a crash.

### CODE REVIEW

File: src/core/layertree/qgslegendpatchshape.cpp [hunk 1]
```diff
@@ -151,10 +151,14 @@ QList<QList<QPolygonF> > QgsLegendPatchShape::toQPolygonF( QgsSymbol::SymbolType
     {
       QList< QList<QPolygonF> > res;
       const QgsGeometry patch = geom;
-      for ( auto it = patch.const_parts_begin(); it != patch.const_parts_end(); ++it )
+      if ( QgsWkbTypes::flatType( mGeometry.wkbType() ) == QgsWkbTypes::LineString || QgsWkbTypes::flatType( mGeometry.wkbType() ) == QgsWkbTypes::MultiLineString)
```
Comment (nyalldawson) [inline src/core/layertree/qgslegendpatchshape.cpp:154]:
```suggestion
      if ( QgsWkbTypes::geometryType( mGeometry.wkbType() ) == QgsWkbTypes::LineGeometry )
```

File: src/core/layertree/qgslegendpatchshape.cpp [hunk 2]
```diff
@@ -151,10 +151,14 @@ QList<QList<QPolygonF> > QgsLegendPatchShape::toQPolygonF( QgsSymbol::SymbolType
...
-      for ( auto it = patch.const_parts_begin(); it != patch.const_parts_end(); ++it )
+      if ( QgsWkbTypes::geometryType( mGeometry.wkbType() ) == QgsWkbTypes::LineGeometry )
       {
-        res << ( QList< QPolygonF >() << curveToPolygonF( qgsgeometry_cast< const QgsCurve * >( *it ) ) );
+        for ( auto it = patch.const_parts_begin(); it != patch.const_parts_end(); ++it )
+        {
+          res << ( QList< QPolygonF >() << curveToPolygonF( qgsgeometry_cast< const QgsCurve * >( *it ) ) );
+        }
       }
+
```
Comment (nyalldawson) [inline src/core/layertree/qgslegendpatchshape.cpp:161]:
```suggestion
```

File: src/core/layertree/qgslegendpatchshape.cpp (hunk 1 above)
Comment (TurboGraphxBeige) [reply src/core/layertree/qgslegendpatchshape.cpp]: Even better!

**Comment by nyalldawson** [comment]
Backported to 3.16 in c2d02311a3504fb655f50e9b609af5a353c92755

**Pull request merged**